### ISSUE TITLE
Clarify handling of invalid codepoint escape sequences

### ISSUE THREAD

**Issue by kasei** [w3c/sparql-query#164]
I think the current spec text is ambiguous about how codepoint escape sequences should be handled if they are invalid. For example:

```
SELECT * WHERE { ?s ?p "\\u000Z" }
```

I think we might want to consider adding (either normative or best-practice) text about how this case should be handled. It seems like several systems (including my own, and Jena) ignore invalid sequences, causing the above query to have a literal that starts with an escaped backslash, followed by the four characters "000Z". Other systems might see the `\u` with invalid trailing characters and raise an error. Having clarity on the expected behavior here would be good.

**Comment by kasei** [comment]
@afs –

> I'm guessing you are [referring to the text](https://www.w3.org/TR/sparql12-query/#codepointEscape) "processed for codepoint escape sequences before parsing"
> 
> What alternative readings do you see?

I think it's unclear what "processed" means here. Should the example I gave me an error? That is, does the simple appearance of `\u` imply a codepoint escape, and the trailing characters being non-HEX mean an error? Or is the sequence only handled as an escape if it matches the 6-character pattern `\u` HEX HEX HEX HEX (similarly for `\U)? *Usually* this difference won't matter because if you don't handle a non-HEX sequence, you'll likely get a syntax error because `\u` isn't going to be valid (e.g. as "Escape sequences in strings"). However, the example above shows that it is possible to get a valid query if you skip the "before parsing" stage, and let the backslash part of `\u` participate as the *escaped character* of a string escape sequence.

> The way Turtle handles this differently to SPARQL - it has `UCHAR` in the grammar and that occurs in strings and URIs. (I think the Turtle way is better but we are where we are.)

Agreed the Turtle handling is better, and also that "we are where we are." So I'm just looking to start a discussion on which of the two possibilities I note above is the expected behavior (if we can find consensus on that), and hoping we can add some text indicating that expectation.

**Comment by afs** [comment]
An alternative is to develop some rdf-tests. The discussion would reach practitioners.

Here are some more examples to add to the collection:

```
SELECT * WHERE { ?s ?p "\\u0041" }
SELECT * WHERE { ?s ?p "\\u0074" }
SELECT * WHERE { ?s ?p "\u005Cn" }
SELECT * WHERE { ?s ?p "\u005C\u005Cn" }
```

Hex x41 is `A` -- `\A` is illegal.
Hex x74 is `t` -- may become a tab (!).
Hex x5C is `\`.

There is an argument SPARQL should switch to Turtle-style on security grounds because of the obfuscation possibilities.
```
\u0041\u0053\u004B\u0020\u007B\u007D
```
which is `ASK{}`

**Comment by kasei** [comment]
Agreed. I can try to work on a PR with some tests in this area (using both approaches) and we could solicit feedback from implementors.

**Comment by kasei** [comment]
w3c/rdf-tests#151 now only contains the tests that I think are unambiguous. I've pulled the following two tests out of that PR, and hope that we can discuss what we want the spec to say about these cases:

```
SELECT * WHERE {
	?s ?p "\\u000Z"
}

# \u000Z is an invalid escape sequence. It should either: be left as-is
# as it does not match the escape syntax of `\u HEX HEX HEX HEX`; or
# seen as a syntax error.
```

```
SELECT * WHERE {
	?s ?p "\\u000\u241e"
}

# \u000\ is an invalid escape sequence, while \u241e is a valid sequence.
# The interpretation here depends on whether the entire six-character sequence `\u000\`
# is consumed during parsing/tokenizing and left as-is (or raised as an error),
# or whether the final character is recognized as the first character of a valid
# codepoint escape sequence. If the latter, the unescaping should result in a
# six-codepoint literal whose five-character prefix is "\u000", followed by
# SYMBOL FOR RECORD SEPARATOR (U+241E).
```

> * What is a codepoint escape sequence? is it something that starts \u/\U or is it only a codepoint escape sequence if it is valid?

My feeling on this is that a codepoint escape sequence should be considered to be the entire 6 byte (or 10 byte, respectively) sequence starting with `\u` (or `\U`). If the trailing 4 (8) bytes are not valid, it should be a syntax error.

> * How is replacement done? Is it left-to-right? What about overlaps? Is it applied repeatedly?

My preference would be left-to-right, which I think implies that overlaps would always be either a syntax error (like `\u012\u0033`) or only considered to be a single escape (in the case the overlap is just on the backslash, as in `\\u0074u0123`).

**Comment by kasei** [comment]
For context here, I think this issue is just waiting on updated spec text (the SPARQL equivalent to the language in w3c/rdf-turtle#145). When that happens, the issues described above should all be resolved.

**Comment by afs** [comment]
There is the text in https://www.w3.org/TR/sparql12-query/#sec-escapes ::
"Escape sequences are processed by taking the sequence of Unicode code points  ..."

**Comment by kasei** [comment]
> There is the text in https://www.w3.org/TR/sparql12-query/#sec-escapes :: "Escape sequences are processed by taking the sequence of Unicode code points ..."

Ah. I must have been looking at an older revision of the draft. After re-reading that section, it does seem that this issue is now fully addressed by the current spec text. Closing.
